IGNITE-13674  Documentation page added for Native Persistence Defragmentation feature (#8465)
apache/ignite: docs/_docs/persistence/native-persistence-defragmentation.adoc

// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
= Persistence Defragmentation

== Overview

> As memory management mechanism of Apache Ignite can only create or reuse pages for user data but never frees them files where Ignite persists data can only grow and never shrinks.

Reviewer: ```suggestion
Apache Ignite memory management mechanism can only create or reuse pages for user data, but it never frees them. So files, where Ignite persists data, can only grow, but never shrink.
```

In most use cases it doesn't cause any problems as once created page can be reused multiple times. However in certain cases it is possible that cache contains very little data but occupies large chunks of disk space because a lot of data was removed from the cache.

> Defragmentation is aimed to enable user to shrink data files and claim back disk space.

Reviewer: ```suggestion
Defragmentation enables user to shrink data files and claim back disk space.
```

[NOTE]
====

> Defragmentation can only be used with historical rebalance enabled (link to historical rebalance page). If historical rebalance is disabled server node always triggers full rebalance after restart throwing away defragmented partition. Full set of data is transferred to the node from other nodes over network, depending of size of data set it may require a lot of time and may slow down the whole cluster as network capacity is important to fulfill user requests.

Reviewer: ```suggestion
Defragmentation can only be used with historical rebalance enabled (link to historical rebalance page). If historical rebalancing is disabled, the server node always triggers full rebalance after the restart, which would throw away the defragmented partition. A full set of data is transferred to the node from other nodes over a network. Depending on the dataset’s size, transferring may require significant time and slow down the whole cluster as network capacity is essential to fulfill user requests.
```

Reviewer: @sergey-chugunov-1985 Please add a link to the **Historical rebalance** page.

Author: I found out that historical rebalancing isn't covered with documentation yet so I removed mention of the link.

> ====

Reviewer: ```suggestion

```

> == Usage: how and when

Reviewer: ```suggestion
== Performing Defragmentation
```

> Defragmentation is costly operation in terms of disk IO so to avoid slowing down user operations it cannot be executed on regular node joined to the cluster. To execute defragmentation user needs to request it first on a particular node or set of nodes and than restart these nodes.

Reviewer: ```suggestion
Defragmentation is a costly operation in terms of disk IO. To avoid slowing down user operations, do not execute it on a regular node joined to the cluster. To perform defragmentation, you need to request it first on a particular node or set of nodes and then restart them.
```

Author: Disagree on the suggested change: "do not execute" implies that it is possible to do the action but is not recommended. In reality executing defragmentation on a regular node is not possible at all. How can we rephrase this to catch this meaning?

Reviewer: Then, lets leave the first version with minor changes (I've also updated the suggested change):

Defragmentation is a costly operation in terms of disk IO. **To avoid slowing down user operations, note that defragmentation cannot be executed on a regular node joined to the cluster.** To perform defragmentation, you need to request it first on a particular node or set of nodes and then restart them.

> To request defragmentation use the following command: *control.(sh|bat) --defragmentation schedule --nodes <consistentIds> [--caches <cacheNames>]*.

Reviewer: ```suggestion
=== Starting Defragmentation

To request defragmentation use the following command: *control.(sh|bat) --defragmentation schedule --nodes <consistentIds> [--caches <cacheNames>]*.
```

Reviewer: Please use a code block for the command here

> After restart node with requested defragmentation will enter special mode called maintenance mode. Node in maintenance doesn't join the rest of the cluster but stays isolated until defragmentation is completed (or cancelled by explicit user request). After that user has to restart the node one more time: it will exit maintenance mode and returns back to normal operations (joins the cluster and starts to serve regular workload).

Reviewer: ```suggestion
After the manual restart, the node with the requested defragmentation enters a special mode called maintenance mode. The node in maintenance mode does not join the rest of the cluster but remains isolated until defragmentation is completed (or canceled by explicit user request). After that, the user has to restart the node one more time: it exits maintenance mode and returns to normal operations (joining the cluster and starting to serve regular workload).
```

[NOTE]
====

> As nodes in maintenance don't participate in serving usual workload, it is not recommended to execute defragmentation on several nodes at once as it reduces number of backups thus increasing the risk of partition loss.

Reviewer: ```suggestion
Nodes in maintenance mode do not participate in serving the regular workload. It is not recommended to execute defragmentation on several nodes simultaneously as it reduces the number of backups, thus increasing the risk of partition loss.
```

====

> When node executes defragmentation it is possible to cancel it using the following command available in control utility: *control.(sh|bat) --defragmentation cancel --host --port.*

Reviewer: Please use a code block for the command here

> 
> When node executes defragmentation it is possible to cancel it using the following command available in control utility: *control.(sh|bat) --defragmentation cancel --host --port.*

Reviewer: ```suggestion

=== Stopping Defragmentation

When a node executes defragmentation, it is possible to cancel it. To stop defragmentation, run the following command available in the control utility: *control.(sh|bat) --defragmentation cancel --host --port.*
```

> For more information about commands refer to their help.

Reviewer: ```suggestion

```

[NOTE]
====

> To reduce disk space requirements during defragmentation caches are defragmented one by one (if defragmentation of more than one cache was requested). To calculate additional space required find the cache that occupies the most disk space. The same amount of disk space is required for defragmentation at max.

Reviewer: ```suggestion
To reduce disk space requirements during defragmentation, caches are defragmented one by one (if a defragmentation of more than one cache is requested). To calculate additional required space, find the cache that occupies the most disk space. The same amount of disk space is required for defragmentation at max.
```

====

== Conclusion

> In most situations defragmentation isn't needed as existing memory management mechanism effectively reuses memory left after data deletion. But in rare cases it may be necessary to employ it to free up disk space.

Reviewer: ```suggestion
In most situations, defragmentation is not necessary as existing memory management mechanism effectively reuses memory left after data deletion. But in rare cases, it may be necessary to free disk space up.
```

> Its usage requires taking nodes out of normal operations so it careful planning is needed.

Reviewer: ```suggestion
Persistence defragmentation requires taking nodes out of their normal operations, so a careful planning is recommended.
```

// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
= Persistence Defragmentation

== Overview

Apache Ignite memory management mechanism can only create or reuse pages for user data, but it never frees them. So the files, where Ignite persists data, can only grow and never shrink.

In most use cases, it does not cause any problem as once created page can be reused multiple times. However, in certain cases, it is possible that cache contains very little data but occupies large chunks of disk space because a significant volume of data was removed from the cache.

> Defragmentation enables user to shrink data files and claim back disk space.

Reviewer: ```suggestion
Defragmentation enables a user to shrink data files and claim back disk space.
```

[NOTE]
====
Defragmentation can only be used with historical rebalance enabled. If historical rebalancing is disabled, the server node always triggers full rebalance after the restart, which would throw away the defragmented partition. A full set of data is transferred to the node from other nodes over a network. Depending on the dataset’s size, transferring may require significant time and slow down the whole cluster as network capacity is essential to fulfill user requests.
====

== Performing Defragmentation

> Defragmentation is costly operation in terms of disk IO. To avoid slowing down user operations, note that defragmentation cannot be executed on a regular node joined to the cluster. To perform defragmentation, you need to request it first on a particular node or set of nodes and then restart them.

Reviewer: ```suggestion
Defragmentation is a costly operation in terms of disk IO. To avoid slowing down user operations, note that defragmentation cannot be executed on a regular node joined to the cluster. To perform defragmentation, you need to request it first on a particular node or set of nodes and then restart them.
```

=== Starting Defragmentation

> To request defragmentation use the following command:

Reviewer: ```suggestion
To request defragmentation, use the following command:
```

[NOTE]
====
Nodes in maintenance mode do not participate in serving the regular workload. It is not recommended to execute defragmentation on several nodes simultaneously as it reduces the number of backups, thus increasing the risk of partition loss.
====

=== Stopping Defragmentation

When a node executes defragmentation, it is possible to cancel it. To stop defragmentation, run the following command available in the control utility:
[source,shell]
----
control.(sh|bat) --defragmentation cancel --host --port
----

[NOTE]
====
To reduce disk space requirements during defragmentation, caches are defragmented one by one (if a defragmentation of more than one cache is requested). To calculate additional required space, find the cache that occupies the most disk space. The same amount of disk space is required for defragmentation at max.
====

== Conclusion

> In most situations defragmentation isn't needed as existing memory management mechanism effectively reuses memory left after data deletion. But in rare cases it may be necessary to employ it to free up disk space.

Reviewer: ```suggestion
In most situations, defragmentation is not necessary as existing memory management mechanism effectively reuses memory left after data deletion. But in rare cases, it may be necessary to free disk space up.
```

> Its usage requires taking nodes out of normal operations so it careful planning is needed.

Reviewer: ```suggestion
Persistence defragmentation requires taking nodes out of their normal operations, so a careful planning is recommended.
```
